File: src/tcp/tcp.cpp
#include "../controller.h"
#include "../coroutine.h"
#include "../udp/udp.h"
#include "tcp.h"
#include "socket.h"
#include "server.h"

namespace flame::tcp {
	static std::unique_ptr<boost::asio::ip::tcp::resolver> resolver_;
	void declare(php::extension_entry& ext)
    {
		gcontroller->on_init([] (const php::array& options) {
            resolver_.reset(new boost::asio::ip::tcp::resolver(gcontroller->context_x));
        })->on_stop([] () {
            resolver_.reset();
        });
		ext
			.function<connect>("flame\\tcp\\connect", {
				{"address", php::TYPE::STRING},
			});
		socket::declare(ext);
		server::declare(ext);
	}

    php::value connect(php::parameters& params) {
		php::object obj(php::class_entry<socket>::entry());
        socket *ptr = static_cast<socket *>(php::native(obj));

        std::string str = params[0];
        auto pair = udp::addr2pair(str);
        if (pair.first.empty() || pair.second.empty()) throw php::exception(zend_ce_type_error
        	, "Failed to connect TCP socket: illegal address format"
        	, -1);

		coroutine_handler ch{coroutine::current};
        boost::system::error_code err;
		boost::asio::ip::tcp::resolver::results_type eps;
		// DNS 地址解析
		resolver_->async_resolve(pair.first, pair.second, [&err, &eps, &ch] (const boost::system::error_code& error, boost::asio::ip::tcp::resolver::results_type results) {
			if (error) err = error;
			else eps = results;
            ch.resume();
		});
		ch.suspend();
		if (err) throw php::exception(zend_ce_exception
            , (boost::format("Failed to resolve address: %s") % err.message()).str()
            , err.value());

		// 连接
        boost::asio::async_connect(ptr->socket_, eps, [&err, &obj, &ptr, &ch](const boost::system::error_code &error, const boost::asio::ip::tcp::endpoint &edp) {
            if (error) err = error;
			else {
	            obj.set("local_address", (boost::format("%s:%d")
	            	% ptr->socket_.local_endpoint().address().to_string()
	            	% ptr->socket_.local_endpoint().port() ).str());
				obj.set("remote_address", (boost::format("%s:%d")
					% ptr->socket_.remote_endpoint().address().to_string()
					% ptr->socket_.remote_endpoint().port() ).str());
			}
            ch.resume();
        });
        ch.suspend();
        if (err) throw php::exception(zend_ce_exception
            , (boost::format("Failed to connect TCP socket: %s") % err.message()).str()
            , err.value());
        return std::move(obj);
	}
} // namespace flame::tcp

File: doc/tcp.php
<?php
namespace flame\tcp;

/**
 * 连接指定地址, 并返回 socket 对象
 * @param string $address 连接地址, 例如: "127.0.0.1:8687"
 */
function connect(string $address):socket {}

class socket {
    public $local_address;
    public $remote_address;
    /**
     * 读取一定量的数据, 实际读取方式根据 $completion 参数决定;
     * @param mixed $completion 
     *  * 参数为 null 时, 表示随机读取一定量数据 (不会超过 8192 字节);
     *  * 参数为 int 时, 表示读取指定量的数据, 例如 $completion = 1024 表示读取 1024 字节;
     *  * 参数为 string 时, 表示读取到指定的结束符, 例如 $completion = "\r\n" 读取到 "abc\r\n" 字符串;
     */
    function read(mixed $completion = null) {}
    /**
     * 发送数据
     * @param string $data
     */
    function write(string $data) {}
    /**
     * 关闭连接
     */
    function close() {}
}
class server {
    /**
     * 创建一个服务器, 绑定在指定的地址上, 准备服务
     * @param string $address 绑定地址, 例如: "0.0.0.0:8888" 或 ":::8888"
     */
    function __construct(string $address) {}
    /**
     * 启动服务器, 当有连接建立时, 启用协程执行回调函数;
     * @param callable $cb 回调函数, 形如:
     *  function($socket) {}
     */
    function run(callable $cb) {}
    /**
     * 关闭服务器
     */
    function close() {}
}